core_v2/sources/v2_domains.move

```diff
@@ -0,0 +1,1315 @@
+        let transfer_ref = &borrow_global_mut<NameRecord>(name_record_address).transfer_ref;
+        if (transferrable) {
+            object::enable_ungated_transfer(transfer_ref);
+        } else {
+            object::disable_ungated_transfer(transfer_ref);
+        }
+    }
+
+    public fun transfer_subdomain_owner(
+        sign: &signer,
+        domain_name: String,
+        subdomain_name: String,
+        new_owner_address: address,
+        new_target_address: Option<address>,
+    ) acquires NameRecord, SubdomainExt, ReverseRecord, SetReverseLookupEvents {
+        // validate user own the domain
+        let signer_addr = signer::address_of(sign);
+        assert!(
+            is_token_owner(signer_addr, domain_name, option::none()),
+            error::permission_denied(ENOT_OWNER_OF_DOMAIN)
+        );
+        assert!(
+            !is_name_expired(domain_name, option::none()),
+            error::permission_denied(ECANNOT_TRANSFER_SUBDOMAIN_WHILE_DOMAIN_HAS_EXPIRED)
+        );
+
+        let token_addr = get_token_addr_inline(domain_name, option::some(subdomain_name));
+        let record = borrow_global_mut<NameRecord>(token_addr);
+        record.target_address = new_target_address;
+        object::transfer_with_ref(object::generate_linear_transfer_ref(&record.transfer_ref), new_owner_address);
+        // clear the primary name
+        clear_reverse_lookup_for_name(option::some(subdomain_name), domain_name);
+    }
+
+    /// this is for domain owner to update subdomain expiration time
+    public fun set_subdomain_expiration(
+        domain_admin: &signer,
+        domain_name: String,
+        subdomain_name: String,
+        expiration_time_sec: u64,
+    ) acquires NameRecord, SubdomainExt {
+        validate_subdomain_registered_and_domain_owned_by_signer(domain_admin, domain_name, subdomain_name);
+        // check if the expiration time is valid
+        let domain_record = get_record(domain_name, option::none());
+        assert!(
+            domain_record.expiration_time_sec >= expiration_time_sec,
+            error::invalid_state(ESUBDOMAIN_EXPIRATION_PASS_DOMAIN_EXPIRATION)
+        );
+
+        // check the auto-renew flag
+        let subdomain_name_opt = option::some(subdomain_name);
+        let token_addr = get_token_addr_inline(domain_name, subdomain_name_opt);
+        let record = borrow_global_mut<NameRecord>(get_token_addr_inline(domain_name, subdomain_name_opt));
+        assert!(exists<SubdomainExt>(token_addr), error::invalid_state(ENOT_A_SUBDOMAIN));
+        let subdomain_ext = borrow_global<SubdomainExt>(token_addr);
+        assert!(
+            subdomain_ext.subdomain_expiration_policy != SUBDOMAIN_POLICY_LOOKUP_DOMAIN_EXPIRATION,
+            error::invalid_state(ESUBDOMAIN_IS_AUTO_RENEW)
+        );
+
+        // manually set the expiration date
+        record.expiration_time_sec = expiration_time_sec;
+    }
+
+    public fun set_subdomain_expiration_policy(
```

Comment:
feels like this isn't the right API

we should probably just have distinct `set_subdomain_expiration_policy_as_xxx` functions